File: sdk/python/src/inferlet_py/brle.py
"""
Binary Run-Length Encoding (BRLE) for attention masks.

This efficiently stores boolean sequences by encoding run lengths.
Mirrors the Rust Brle from inferlet/src/brle.rs and JS from inferlet-js/src/brle.ts.
"""

from __future__ import annotations


class Brle:
    """
    Binary Run-Length Encoding for attention masks.

    Encodes boolean sequences as alternating run lengths of false/true.
    In attention mask context: false = visible (can attend).

    Examples:
        [false, false, true, true, true, false] → [2, 3, 1]
        [true, true, false] → [0, 2, 1]
    """

    def __init__(self, buffer: list[int], total_size: int) -> None:
        self.buffer = buffer
        self.total_size = total_size

    @classmethod
    def new(cls, size: int) -> "Brle":
        """
        Create a Brle with `size` visible positions (all can attend).

        In attention mask semantics: false = visible.
        """
        if size == 0:
            return cls([], 0)
        # All false values = [size] means size visible positions
        return cls([size], size)

    @classmethod
    def from_array(cls, values: list[bool]) -> "Brle":
        """Create a Brle from an array of booleans."""
        if not values:
            return cls.new(0)

        buffer: list[int] = []
        current_val = False
        count = 0

        # Handle initial run of false
        if values[0]:
            buffer.append(0)
            current_val = True

        for val in values:
            if val == current_val:
                count += 1
            else:
                buffer.append(count)
                current_val = val
                count = 1

        buffer.append(count)
        return cls(buffer, len(values))

    def __len__(self) -> int:
        return self.total_size

    def clone(self) -> "Brle":
        """Create a copy of this Brle."""
        return Brle(self.buffer.copy(), self.total_size)

    def append(self, value: bool) -> None:
        """Append a single boolean value."""
        if self.total_size == 0:
            if value:
                self.buffer = [0, 1]
            else:
                self.buffer = [1]
            self.total_size = 1
            return

        self.total_size += 1
        is_last_run_true = len(self.buffer) % 2 == 0

        if value == is_last_run_true:
            self.buffer[-1] += 1
        else:
            self.buffer.append(1)

    def truncate(self, new_size: int) -> "Brle":
        """Return a new Brle truncated to new_size."""
        if new_size >= self.total_size:
            return self.clone()
        if new_size <= 0:
            return Brle.new(0)

        new_buffer: list[int] = []
        remaining = new_size

        for run_len in self.buffer:
            if remaining <= 0:
                break
            if run_len <= remaining:
                new_buffer.append(run_len)
                remaining -= run_len
            else:
                new_buffer.append(remaining)
                remaining = 0

        return Brle(new_buffer, new_size)

    def _slice(self, start: int, end: int) -> "Brle":
        """
        Create a new Brle representing a slice [start, end) of this one.

        Internal method used by remove_range.
        """
        end = min(end, self.total_size)
        if start >= end:
            return Brle.new(0)

        new_size = end - start
        new_buffer: list[int] = []

        current_pos = 0
        for i, run_len in enumerate(self.buffer):
            value = i % 2 != 0  # Even indices are false, odd are true
            r_start = current_pos
            r_end = current_pos + run_len
            current_pos = r_end

            # Calculate intersection of [r_start, r_end) and [start, end)
            slice_r_start = max(r_start, start)
            slice_r_end = min(r_end, end)

            if slice_r_start < slice_r_end:
                length = slice_r_end - slice_r_start

                if not new_buffer:
                    # First run of the new slice
                    if value:
                        new_buffer.append(0)  # Start with zero-length false run
                    new_buffer.append(length)
                else:
                    # Check if we can merge with previous run
                    last_run_is_true = (len(new_buffer) - 1) % 2 != 0
                    if last_run_is_true == value:
                        new_buffer[-1] += length
                    else:
                        new_buffer.append(length)

        return Brle(new_buffer, new_size)

    def remove_range(self, start: int, end: int) -> None:
        """
        Remove a range of boolean values [start, end).

        Modifies this Brle in place by concatenating the parts before
        and after the removed range.
        """
        end = min(end, self.total_size)
        if start >= end:
            return

        head = self._slice(0, start)
        tail = self._slice(end, self.total_size)

        # Merge head and tail
        if head.total_size == 0:
            self.buffer = tail.buffer.copy()
            self.total_size = tail.total_size
        elif tail.total_size == 0:
            self.buffer = head.buffer.copy()
            self.total_size = head.total_size
        else:
            # Check if last run of head and first run of tail have same value
            head_last_is_true = (len(head.buffer) - 1) % 2 != 0
            tail_first_is_true = tail.buffer[0] == 0 and len(tail.buffer) > 1

            if head_last_is_true == tail_first_is_true:
                # Merge the runs
                tail_first_run_len = tail.buffer[1] if tail_first_is_true else tail.buffer[0]
                tail_slice_start = 2 if tail_first_is_true else 1

                new_buffer = head.buffer.copy()
                new_buffer[-1] += tail_first_run_len
                new_buffer.extend(tail.buffer[tail_slice_start:])
                self.buffer = new_buffer
            else:
                # No merge needed
                if tail_first_is_true:
                    self.buffer = head.buffer + tail.buffer[1:]
                else:
                    self.buffer = head.buffer + tail.buffer

            self.total_size = head.total_size + tail.total_size


    def is_empty(self) -> bool:
        """Check if the sequence is empty."""
        return self.total_size == 0

    def iter_runs(self):
        """
        Iterate over the runs, yielding (value, start_index, end_index).

        Yields:
            Tuples of (bool, int, int) representing (value, start, end)
            where value is the boolean value of the run.
        """
        current_pos = 0
        for i, run_len in enumerate(self.buffer):
            value = i % 2 != 0  # Even indices are false, odd are true
            start = current_pos
            end = current_pos + run_len
            current_pos = end
            yield (value, start, end)

    def is_masked(self, indices: list[int]) -> list[bool]:
        """
        Check the boolean values at a given set of indices.

        Returns whether each index is masked (TRUE = masked/hidden from attention).
        This is optimized to check multiple indices in a single pass over the data.

        Args:
            indices: A list of indices to check. Does not need to be sorted.

        Returns:
            A list of booleans containing the value at each corresponding index.

        Raises:
            IndexError: If any index is out of bounds.
        """
        if not indices:
            return []

        # Pair indices with their original position for result ordering
        indexed_indices = [(i, idx) for i, idx in enumerate(indices)]
        indexed_indices.sort(key=lambda x: x[1])

        results = [False] * len(indices)
        run_iter = iter(self.iter_runs())
        current_run = next(run_iter, None)

        for original_pos, query_index in indexed_indices:
            if query_index >= self.total_size:
                raise IndexError(
                    f"Index {query_index} is out of bounds for Brle of length {self.total_size}"
                )

            # Advance through runs until we find the one containing the query_index
            while current_run is not None:
                value, run_start, run_end = current_run
                if query_index >= run_start and query_index < run_end:
                    results[original_pos] = value
                    break
                current_run = next(run_iter, None)

        return results

    def is_range_all_value(self, start: int, end: int, expected_value: bool) -> bool:
        """
        Check if all boolean values within a specified range [start, end)
        are equal to a given expected_value.

        Args:
            start: The starting index of the range (inclusive).
            end: The ending index of the range (exclusive).
            expected_value: The boolean value to check for.

        Returns:
            True if every value in the range is expected_value, False otherwise.
            Returns True for an empty range (start >= end).
        """
        if start >= end:
            return True
        if end > self.total_size:
            return False

        # Track how much of the target range we've verified
        pos_covered = start

        for run_value, run_start, run_end in self.iter_runs():
            # Find intersection of current run [run_start, run_end) and [pos_covered, end)
            intersect_start = max(run_start, pos_covered)
            intersect_end = min(run_end, end)

            if intersect_start < intersect_end:
                # There's a valid intersection - check if value matches
                if run_value != expected_value:
                    return False
                # Update coverage marker
                pos_covered = intersect_end

                # If we've covered the entire target range, we're done
                if pos_covered >= end:
                    return True

            # Optimization: if current run extends beyond target, no need to check more
            if run_end >= end:
                break

        return pos_covered >= end

    def to_array(self) -> list[bool]:
        """
        Convert to a list of booleans.

        Returns:
            List of boolean values represented by this Brle
        """
        result: list[bool] = []
        current_pos = 0

        for i, run_len in enumerate(self.buffer):
            value = i % 2 != 0  # Even = false, odd = true
            result.extend([value] * run_len)
            current_pos += run_len

        return result

    def mask(self, indices: list[int], value: bool) -> None:
        """
        Set specific indices to the given value.

        Args:
            indices: Indices to modify
            value: Value to set (True = masked, False = visible)
        """
        if not indices:
            return

        # Sort and deduplicate
        sorted_indices = sorted(set(indices))

        # Group into contiguous ranges
        ranges: list[tuple[int, int]] = []
        range_start = sorted_indices[0]
        range_end = range_start + 1

        for idx in sorted_indices[1:]:
            if idx == range_end:
                range_end = idx + 1
            else:
                ranges.append((range_start, range_end))
                range_start = idx
                range_end = idx + 1
        ranges.append((range_start, range_end))

        # Apply each range
        for start, end in ranges:
            self.mask_range(start, end, value)

    def mask_range(self, start: int, end: int, value: bool) -> None:
        """
        Set a range of indices to the given value.

        Args:
            start: Start index (inclusive)
            end: End index (exclusive)
            value: Value to set (True = masked, False = visible)
        """
        if start >= end or start >= self.total_size:
            return

        end = min(end, self.total_size)

        # Rebuild the buffer with the range modified
        new_buffer: list[int] = []
        current_pos = 0

        for i, run_len in enumerate(self.buffer):
            is_true = i % 2 != 0
            r_start = current_pos
            r_end = current_pos + run_len
            current_pos = r_end

            # Check overlap with [start, end)
            overlap_start = max(r_start, start)
            overlap_end = min(r_end, end)

            if overlap_start >= overlap_end:
                # No overlap - keep original run
                self._append_run(new_buffer, is_true, run_len)
            else:
                # Has overlap - split into up to 3 parts
                # Part 1: before overlap
                if r_start < overlap_start:
                    self._append_run(new_buffer, is_true, overlap_start - r_start)

                # Part 2: the overlap (with new value)
                self._append_run(new_buffer, value, overlap_end - overlap_start)

                # Part 3: after overlap
                if overlap_end < r_end:
                    self._append_run(new_buffer, is_true, r_end - overlap_end)

        self.buffer = new_buffer

    @staticmethod
    def _append_run(buffer: list[int], value: bool, length: int) -> None:
        """
        Helper to append a run to a buffer, merging if possible.

        Args:
            buffer: The buffer to append to
            value: The boolean value of the run
            length: The length of the run
        """
        if length <= 0:
            return

        if not buffer:
            if value:
                buffer.extend([0, length])
            else:
                buffer.append(length)
        else:
            last_is_true = (len(buffer) - 1) % 2 != 0
            if last_is_true == value:
                buffer[-1] += length
            else:
                buffer.append(length)

    def extend(self, other: "Brle") -> None:
        """
        Extend this Brle with another Brle's contents.

        Args:
            other: Brle to append
        """
        if other.total_size == 0:
            return

        if self.total_size == 0:
            self.buffer = other.buffer.copy()
            self.total_size = other.total_size
            return

        # Check if last run of self and first run of other have same value
        self_last_is_true = (len(self.buffer) - 1) % 2 != 0
        other_first_is_true = other.buffer[0] == 0 and len(other.buffer) > 1

        if self_last_is_true == other_first_is_true:
            # Merge the runs
            first_run_len = other.buffer[1] if other_first_is_true else other.buffer[0]
            slice_start = 2 if other_first_is_true else 1

            self.buffer[-1] += first_run_len
            self.buffer.extend(other.buffer[slice_start:])
        else:
            # No merge - just concatenate
            if other_first_is_true:
                self.buffer.extend(other.buffer[1:])
            else:
                self.buffer.extend(other.buffer)

        self.total_size += other.total_size


def causal_mask(num_total_tokens: int, num_input_tokens: int) -> list[Brle]:
    """
    Create causal attention masks for the given parameters.

    Each token can only attend to tokens at positions <= its own position.
    For prefill, this means progressively longer masks covering full kv_len.

    BRLE format: [visible_count, masked_count] where:
    - visible (false) = can attend to these positions
    - masked (true) = cannot attend to these positions

    Args:
        num_total_tokens: Total number of tokens in the KV cache after processing
        num_input_tokens: Number of new input tokens being processed

    Returns:
        List of Brle masks, one per input token, each of length num_total_tokens
    """
    masks: list[Brle] = []
    offset = num_total_tokens - num_input_tokens

    for i in range(num_input_tokens):
        # Token at position (offset + i) can attend to positions 0..(offset + i) inclusive
        visible = offset + i + 1
        masked = num_total_tokens - visible

        if masked > 0:
            # [visible_count, masked_count]
            masks.append(Brle([visible, masked], num_total_tokens))
        else:
            # Last token can see everything - just [visible_count]
            masks.append(Brle([visible], num_total_tokens))

    return masks


def causal_mask_raw(num_total_tokens: int, num_input_tokens: int) -> list[list[int]]:
    """
    Create causal attention masks as raw buffers (for WIT interface).

    Args:
        num_total_tokens: Total number of tokens in the KV cache after processing
        num_input_tokens: Number of new input tokens being processed

    Returns:
        List of raw BRLE buffers (list[list[int]]) for passing to WIT
    """
    masks = causal_mask(num_total_tokens, num_input_tokens)
    return [m.buffer for m in masks]

File: sdk/javascript/src/brle.ts
// Binary Run-Length Encoding (BRLE) structure.
// This efficiently stores a large sequence of booleans by encoding
// the lengths of consecutive runs of `false` and `true` values.
// Mirrors the Rust Brle from inferlet/src/brle.rs

/**
 * A Binary Run-Length Encoding (BRLE) structure.
 *
 * This structure efficiently stores a large sequence of booleans by encoding
 * the lengths of consecutive runs of `false` and `true` values.
 * The sequence always begins with a run of `false`s, which may be zero-length.
 *
 * Encoding examples:
 * - `[false, false, true, true, true, false]` is encoded as `[2, 3, 1]`.
 * - `[true, true, false]` is encoded as `[0, 2, 1]`.
 */
export class Brle {
  /** The buffer of run lengths. Even indices are for `false` runs, odd for `true`. */
  buffer: number[];
  /** The total number of boolean values represented. */
  totalSize: number;

  private constructor(buffer: number[], totalSize: number) {
    this.buffer = buffer;
    this.totalSize = totalSize;
  }

  /**
   * Creates a new `Brle` instance representing `size` `false` values.
   * Matches Rust semantics where FALSE = visible (can attend).
   */
  static new(size: number): Brle {
    if (size === 0) {
      return new Brle([], 0);
    }
    // All false values: [size] means size false values (visible positions)
    return new Brle([size], size);
  }

  /**
   * Creates a `Brle` from an array of booleans.
   */
  static fromArray(values: boolean[]): Brle {
    if (values.length === 0) {
      return Brle.new(0);
    }

    const buffer: number[] = [];
    let currentVal = false;
    let count = 0;

    // Handle the initial run of `false`s, which could be zero-length.
    if (values[0]) {
      buffer.push(0);
      currentVal = true;
    }

    for (const val of values) {
      if (val === currentVal) {
        count++;
      } else {
        buffer.push(count);
        currentVal = val;
        count = 1;
      }
    }
    buffer.push(count); // Push the last run

    return new Brle(buffer, values.length);
  }

  /**
   * Returns the total number of booleans in the sequence.
   */
  len(): number {
    return this.totalSize;
  }

  /**
   * Returns `true` if the sequence is empty.
   */
  isEmpty(): boolean {
    return this.totalSize === 0;
  }

  /**
   * Decodes the `Brle` into an array of booleans.
   */
  toArray(): boolean[] {
    const result: boolean[] = [];
    for (const [value, start, end] of this.iterRuns()) {
      const runLen = end - start;
      for (let i = 0; i < runLen; i++) {
        result.push(value);
      }
    }
    return result;
  }

  /**
   * Appends a boolean value to the end of the sequence.
   */
  append(flag: boolean): void {
    if (this.buffer.length === 0) {
      if (flag) {
        this.buffer.push(0, 1);
      } else {
        this.buffer.push(1);
      }
    } else {
      const lastRunIsTrue = (this.buffer.length - 1) % 2 !== 0;
      if (lastRunIsTrue === flag) {
        this.buffer[this.buffer.length - 1]++;
      } else {
        this.buffer.push(1);
      }
    }
    this.totalSize++;
  }

  /**
   * Sets a range of booleans to a specified value.
   * The range is exclusive of `end` (`start..end`).
   */
  maskRange(start: number, end: number, flag: boolean): void {
    if (start >= end) {
      return;
    }
    this.maskInternal([[start, end]], flag);
  }

  /**
   * Sets multiple indices to a specified value.
   */
  mask(indices: number[], flag: boolean): void {
    if (indices.length === 0) {
      return;
    }

    // Sort and deduplicate indices
    const sortedIndices = [...new Set(indices)].sort((a, b) => a - b);

    // Group into contiguous ranges
    const ranges: [number, number][] = [];
    let rangeStart = sortedIndices[0];
    let rangeEnd = rangeStart + 1;

    for (let i = 1; i < sortedIndices.length; i++) {
      const index = sortedIndices[i];
      if (index === rangeEnd) {
        rangeEnd = index + 1;
      } else {
        ranges.push([rangeStart, rangeEnd]);
        rangeStart = index;
        rangeEnd = index + 1;
      }
    }
    ranges.push([rangeStart, rangeEnd]);

    this.maskInternal(ranges, flag);
  }

  /**
   * Returns an iterator over the runs, yielding [value, startIndex, endIndex].
   */
  *iterRuns(): Generator<[boolean, number, number]> {
    let currentPos = 0;
    for (let i = 0; i < this.buffer.length; i++) {
      const runLen = this.buffer[i];
      const value = i % 2 !== 0;
      const start = currentPos;
      const end = currentPos + runLen;
      currentPos = end;
      yield [value, start, end];
    }
  }

  /**
   * Checks the boolean values at a given set of indices.
   * Returns whether each index is masked (TRUE = masked/hidden from attention).
   *
   * This is optimized to check multiple indices in a single pass over the data.
   *
   * @param indices - An array of indices to check. The indices do not need to be sorted.
   * @returns An array of booleans containing the value at each corresponding index.
   */
  isMasked(indices: number[]): boolean[] {
    if (indices.length === 0) {
      return [];
    }

    // Pair indices with their original position for result ordering
    const indexedIndices: [number, number][] = indices.map((idx, i) => [i, idx]);
    indexedIndices.sort((a, b) => a[1] - b[1]);

    const results = new Array<boolean>(indices.length).fill(false);
    const runIterator = this.iterRuns();
    let currentRun = runIterator.next();

    for (const [originalPos, queryIndex] of indexedIndices) {
      if (queryIndex >= this.totalSize) {
        throw new Error(
          `Index ${queryIndex} is out of bounds for Brle of length ${this.totalSize}`
        );
      }

      // Advance through runs until we find the one containing the queryIndex
      while (!currentRun.done) {
        const [value, runStart, runEnd] = currentRun.value;
        if (queryIndex >= runStart && queryIndex < runEnd) {
          results[originalPos] = value;
          break;
        }
        currentRun = runIterator.next();
      }
    }

    return results;
  }

  /**
   * Checks if all boolean values within a specified range [start, end)
   * are equal to a given expectedValue.
   *
   * @param start - The starting index of the range (inclusive).
   * @param end - The ending index of the range (exclusive).
   * @param expectedValue - The boolean value to check for.
   * @returns true if every value in the range is expectedValue, false otherwise.
   *          Returns true for an empty range (start >= end).
   */
  isRangeAllValue(start: number, end: number, expectedValue: boolean): boolean {
    if (start >= end) {
      return true;
    }
    if (end > this.totalSize) {
      return false;
    }

    // Track how much of the target range we've verified
    let posCovered = start;

    for (const [runValue, runStart, runEnd] of this.iterRuns()) {
      // Find intersection of current run [runStart, runEnd) and [posCovered, end)
      const intersectStart = Math.max(runStart, posCovered);
      const intersectEnd = Math.min(runEnd, end);

      if (intersectStart < intersectEnd) {
        // There's a valid intersection - check if value matches
        if (runValue !== expectedValue) {
          return false;
        }
        // Update coverage marker
        posCovered = intersectEnd;

        // If we've covered the entire target range, we're done
        if (posCovered >= end) {
          return true;
        }
      }

      // Optimization: if current run extends beyond target, no need to check more
      if (runEnd >= end) {
        break;
      }
    }

    return posCovered >= end;
  }

  /**
   * Extends this Brle with another one by concatenating their sequences.
   *
   * @param other - The Brle to append to this one.
   */
  extend(other: Brle): void {
    if (other.isEmpty()) {
      return;
    }
    if (this.isEmpty()) {
      this.buffer = [...other.buffer];
      this.totalSize = other.totalSize;
      return;
    }

    const selfLastRunIsTrue = (this.buffer.length - 1) % 2 !== 0;
    const otherFirstRunIsTrue = other.buffer[0] === 0 && other.buffer.length > 1;

    if (selfLastRunIsTrue === otherFirstRunIsTrue) {
      // Merge the last run of self with the first run of other
      const otherFirstRunLen = otherFirstRunIsTrue
        ? other.buffer[1]
        : other.buffer[0];
      const otherSliceStart = otherFirstRunIsTrue ? 2 : 1;

      this.buffer[this.buffer.length - 1] += otherFirstRunLen;
      this.buffer.push(...other.buffer.slice(otherSliceStart));
    } else {
      // No merge needed - skip zero-length false run if present
      if (otherFirstRunIsTrue) {
        this.buffer.push(...other.buffer.slice(1));
      } else {
        this.buffer.push(...other.buffer);
      }
    }
    this.totalSize += other.totalSize;
  }

  /**
   * Core masking logic for pre-sorted, disjoint ranges.
   */
  private maskInternal(ranges: [number, number][], flag: boolean): void {
    if (ranges.length === 0 || this.totalSize === 0) {
      return;
    }

    // Collect all event points
    const events = new Set<number>();
    events.add(0);
    events.add(this.totalSize);

    for (const [start, end] of ranges) {
      const clampedStart = Math.min(start, this.totalSize);
      const clampedEnd = Math.min(end, this.totalSize);
      if (clampedStart < clampedEnd) {
        events.add(clampedStart);
        events.add(clampedEnd);
      }
    }

    for (const [, runStart, runEnd] of this.iterRuns()) {
      events.add(runStart);
      events.add(runEnd);
    }

    const eventPoints = [...events].sort((a, b) => a - b);
    const newBuffer: number[] = [];
    const runIterator = this.iterRuns();
    let rangeIndex = 0;
    let currentRun = runIterator.next();

    for (let i = 0; i < eventPoints.length - 1; i++) {
      const start = eventPoints[i];
      const end = eventPoints[i + 1];
      if (start >= end) continue;

      const midPoint = start + Math.floor((end - start) / 2);

      // Check if midPoint is in a masked range
      while (rangeIndex < ranges.length && midPoint >= ranges[rangeIndex][1]) {
        rangeIndex++;
      }
      const isMasked =
        rangeIndex < ranges.length &&
        midPoint >= ranges[rangeIndex][0] &&
        midPoint < ranges[rangeIndex][1];

      let value: boolean;
      if (isMasked) {
        value = flag;
      } else {
        // Find the original value
        while (!currentRun.done && midPoint >= currentRun.value[2]) {
          currentRun = runIterator.next();
        }
        value = currentRun.value![0];
      }

      const len = end - start;

      // Merge with previous run if same value
      const shouldMerge =
        newBuffer.length > 0 &&
        (newBuffer.length - 1) % 2 !== 0 === value;

      if (shouldMerge) {
        newBuffer[newBuffer.length - 1] += len;
      } else {
        if (newBuffer.length === 0 && value) {
          newBuffer.push(0);
        }
        newBuffer.push(len);
      }
    }

    this.buffer = newBuffer;
  }

  /**
   * Creates a clone of this Brle.
   */
  clone(): Brle {
    return new Brle([...this.buffer], this.totalSize);
  }

  /**
   * Removes a range of boolean values. The range is exclusive (`start..end`).
   * Creates a new Brle by concatenating the parts before and after the range.
   */
  removeRange(start: number, end: number): void {
    const clampedEnd = Math.min(end, this.totalSize);
    if (start >= clampedEnd) {
      return;
    }

    const head = this.slice(0, start);
    const tail = this.slice(clampedEnd, this.totalSize);

    // Merge head and tail
    if (head.totalSize === 0) {
      this.buffer = [...tail.buffer];
      this.totalSize = tail.totalSize;
    } else if (tail.totalSize === 0) {
      this.buffer = [...head.buffer];
      this.totalSize = head.totalSize;
    } else {
      // Need to properly merge - check if last run of head and first run of tail have same value
      const headLastIsTrue = (head.buffer.length - 1) % 2 !== 0;
      const tailFirstIsTrue = tail.buffer[0] === 0 && tail.buffer.length > 1;

      if (headLastIsTrue === tailFirstIsTrue) {
        // Merge the runs
        const tailFirstRunLen = tailFirstIsTrue ? tail.buffer[1] : tail.buffer[0];
        const tailSliceStart = tailFirstIsTrue ? 2 : 1;

        const newBuffer = [...head.buffer];
        newBuffer[newBuffer.length - 1] += tailFirstRunLen;
        newBuffer.push(...tail.buffer.slice(tailSliceStart));

        this.buffer = newBuffer;
      } else {
        // No merge needed
        if (tailFirstIsTrue) {
          this.buffer = [...head.buffer, ...tail.buffer.slice(1)];
        } else {
          this.buffer = [...head.buffer, ...tail.buffer];
        }
      }
      this.totalSize = head.totalSize + tail.totalSize;
    }
  }

  /**
   * Truncates the Brle to the first `newSize` elements.
   * If newSize >= totalSize, returns a clone unchanged.
   */
  truncate(newSize: number): Brle {
    if (newSize >= this.totalSize) {
      return this.clone();
    }
    return this.slice(0, newSize);
  }

  /**
   * Creates a new Brle representing a slice of the current one.
   */
  private slice(start: number, end: number): Brle {
    const clampedEnd = Math.min(end, this.totalSize);
    if (start >= clampedEnd) {
      return Brle.new(0);
    }

    const newSize = clampedEnd - start;
    const newBuffer: number[] = [];

    for (const [val, rStart, rEnd] of this.iterRuns()) {
      const sliceRStart = Math.max(rStart, start);
      const sliceREnd = Math.min(rEnd, clampedEnd);

      if (sliceRStart < sliceREnd) {
        const len = sliceREnd - sliceRStart;

        if (newBuffer.length === 0) {
          if (val) {
            newBuffer.push(0);
          }
          newBuffer.push(len);
        } else {
          const lastRunIsTrue = (newBuffer.length - 1) % 2 !== 0;
          if (lastRunIsTrue === val) {
            newBuffer[newBuffer.length - 1] += len;
          } else {
            newBuffer.push(len);
          }
        }
      }
    }

    return new Brle(newBuffer, newSize);
  }
}

/**
 * Creates a causal attention mask for the given parameters.
 * Each token can only attend to tokens at positions <= its own position.
 *
 * @param numTotalTokens - Total number of tokens in the KV cache
 * @param numInputTokens - Number of new input tokens being processed
 * @returns Array of Brle masks, one per input token
 */
export function causalMask(numTotalTokens: number, numInputTokens: number): Brle[] {
  const masks: Brle[] = [];
  const offset = numTotalTokens - numInputTokens;
  for (let i = 0; i < numInputTokens; i++) {
    masks.push(Brle.new(offset + i + 1));
  }
  return masks;
}
